enhydris_api_client/__init__.py

```diff
@@ -22,95 +16,94 @@ def urljoin(*args):
-            # See comment about x-www-form-urlencoded above
-        },
-        data="username={}&password={}".format(username, password),
-        cookies=r.cookies,
-        allow_redirects=False,
-    )
-    r1.raise_for_status()
-    result = r1.cookies
-    return result
-
-
-def get_model(base_url, session_cookies, model, id):
-    url = urljoin(base_url, "api/{}/{}/".format(model, id))
-    r = requests.get(url, cookies=session_cookies)
-    r.raise_for_status()
-    return r.json()
-
-
-def post_model(base_url, session_cookies, model, data):
-    r = requests.post(
-        urljoin(base_url, "api/{}/".format(model)),
-        headers={
-            "X-CSRFToken": session_cookies["csrftoken"],
-            "Content-Type": "application/x-www-form-urlencoded"
-            # See comment about x-www-form-urlencoded above
-        },
-        cookies=session_cookies,
-        data=data,
-    )
-    r.raise_for_status()
-    return r.json()["id"]
-
-
-def delete_model(base_url, session_cookies, model, id):
-    url = urljoin(base_url, "api/{}/{}/".format(model, id))
-    r = requests.delete(
-        url,
-        cookies=session_cookies,
-        headers={"X-CSRFToken": session_cookies["csrftoken"]},
-    )
-    if r.status_code != 204:
-        raise requests.exceptions.HTTPError()
-
-
-def read_tsdata(base_url, session_cookies, ts_id):
-    url = urljoin(base_url, "api/tsdata/{0}/".format(ts_id))
-    r = requests.get(url, cookies=session_cookies)
-    r.raise_for_status()
-    return pd.read_csv(StringIO(r.text), header=None, parse_dates=True, index_col=0)
-
-
-def post_tsdata(base_url, session_cookies, timeseries_id, ts):
-    f = StringIO()
-    ts.to_csv(f, header=False)
-    url = urljoin(base_url, "api/tsdata/{}/".format(timeseries_id))
-    r = requests.post(
-        url,
-        data={"timeseries_records": f.getvalue()},
-        headers={
-            "Content-Type": "application/x-www-form-urlencoded",
-            "X-CSRFToken": session_cookies["csrftoken"],
-        },
-        cookies=session_cookies,
-    )
-    r.raise_for_status()
-    return r.text
-
-
-def get_ts_end_date(base_url, session_cookies, ts_id):
-    url = urljoin(base_url, "timeseries/d/{}/bottom/".format(ts_id))
-    r = requests.get(url, cookies=session_cookies)
-    r.raise_for_status()
-    lines = r.text.splitlines()
-    lines.reverse()
-    for line in [x.strip() for x in lines]:
-        if not line:
-            continue
-        datestring = line.split(",")[0]
-        return iso8601.parse_date(datestring, default_timezone=None)
-    return datetime(1, 1, 1)
+class EnhydrisApiClient:
+    def __init__(self, base_url):
+        self.base_url = base_url
+        self.cookies = {}
+
+    def login(self, username, password):
+        self.cookies = {}
+        if not username:
+            return
+        login_url = urljoin(self.base_url, "accounts/login/")
+        r = requests.get(login_url)
+        r.raise_for_status()
+        r1 = requests.post(
```

Comment:
I dont understand why you are using this double / cookie based / complicated solution. 

My conceptual steps: 
- Enhydris app (which this client is made) uses already the `rest_framework`  
- Token and Session Authentication can be easy use out of the box using `rest_framework` ( docs https://www.django-rest-framework.org/api-guide/authentication/ and example: https://github.com/kickapoo/blogapi/commit/1ca6c74384a2c16c538ac8f3107d46f22c982cff)
- With the installation  of `django-rest-auth` (https://github.com/Tivix/django-rest-auth) package in Enydris you can easy have all the rest-auth endpoints (example https://github.com/kickapoo/blogapi/commit/94f2fdf8e4bb853d5f2ac88f74a89c802a41edb2) 
 - With the above you can remove this ugly cookies because basically you are simulating the browser, `r` is the get login url and `r1` is posting the login form. 
- Finally you even use the session of Request if Token is not prefer and only a session is suitable (Token means that you add a new Table that needs to be populate for the Enhydris users)

```python
# example
class EnhydrisApiClient(object):
     def __init__(self, ..):
        self.session = requests.Session()
        self.login(username, password)

   def login(self, username, password):
       r = self.session.post(..login url..., data={'username': .. })
```